api/types/network: generate network-inspect struct definitions from Swagger spec

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- Replaced the hand-written struct definitions of `network.Inspect`, `network.Summary` and some of the referenced struct definitions with functionally-equivalent ones generated from the Swagger spec.
- Added some missing network definitions to the Swagger spec.

**- How I did it**
I had to fork another go-swagger template to work around a curious behaviour where go-swagger would always force the "v" in `IPv4` or `IPv6` to uppercase. [It's not supposed to do that.](https://github.com/go-openapi/swag/blob/bf60a5a89f1f1b17352bd38f681739f06741ccd8/mangling/initialism_index.go#L86-L87)

Convincing go-swagger to generate struct definitions with non-pointer struct-valued fields is not too difficult once you get the hang of it. The trick is annotating the definition of the referenced type with `x-omitempty: false` and `x-nullable: false`. Passing `--keep-spec-order` allows us to maintain the exact field order as the hand-rolled definitions [to make them 100% drop-in replacements.](https://github.com/moby/moby/blob/3eceb2d0e98bd24ad00063744d50e50530243b69/daemon/network.go#L704)

**- How to verify it**
By inspection. The new structs should be identical to the hand-written structs aside from doc comment formatting and the presence of `json` struct-tags that explicitly set the field names to the `encoding/json` defaults. The struct field names and types should all match up 1:1 to the old struct definitions.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
 - Swagger definitions for `NetworkSummary` and `NetworkInspect` have been added to the Swagger spec describing the Engine API.
```

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
api/t/network: generate Inspect, Summary structs

Replace the hand-rolled Network, Summary and Inspect struct types in
api/types/network with types generated from the Swagger definition.

Disable the generation of all unwanted marshalers and unmarshalers.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/templates: vendor structfield template

Taken verbatim from
https://github.com/go-swagger/go-swagger/blob/eee6eaf67f2f342970d277c85ba73f05aed4d114/generator/templates/structfield.gotmpl
so the alterations from the upstream template can be easily diffed.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on api/types/network/network.go:
```diff
@@ -40,32 +38,10 @@ type CreateRequest struct {
-// for [Inspect], but may diverge in the future, as not all information may
-// be included when listing networks.
-type Summary = Inspect
+// Inspect is the body of the "get network" http response message. It currently
+// is an alias for [Summary], but may diverge in the future, as not all
+// information may be included when listing networks.
+type Inspect = Summary
```
Do you think we should make them both concrete types? (and switching to the API module to be a good opportunity to do so?)

thaJeztah replied:
Yeah, I think my intent with the alias was indeed to make them concrete types; sharing the types has been problematic in various cases where we (probably) were too lazy to define a new type, and just reused an existing type, leaving unused fields unset.

I'm not _entirely_ sure if we should embed the type (i.e. `Inspect` _must be_ a superset of `Summary` - it may be possible that fields could make sense for `Summary` that are a different presentation of fields in `Inspect`

corhere replied:
> it may be possible that fields could make sense for `Summary` that are a different presentation of fields in `Inspect`

Serving both presentations could save a round-trip for clients that need both for whatever reason?

I take the fact `Inspect` and `Summary` have been sharing a type definition all this time as a strong hint that there is a common set of properties that any API response describing a network should be including. What do you think of doing something more along the lines of this?
```go
type Summary struct {...}

type Inspect struct {
        Summary
        ...
}

type NetworkListResponseItem struct {
        Summary
        ...
}
```

corhere committed:
api/t/network: generate more structs from Swagger

Replace hand-rolled struct definitions for api/types/network with
types generated from the Swagger definitions:
  - ConfigReference
  - EndpointResource
  - NetworkingConfig
  - PeerInfo
  - ServiceInfo
  - Task

Add Swagger definitions for ServiceInfo and Task.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on api/types/network/summary.go:
```diff
@@ -0,0 +1,41 @@
+	if err != nil {
+		return nil, err
+	}
+	_parts = append(_parts, aO0)
+	return swag.ConcatJSON(_parts...), nil
```
Ah, I see now what you meant!

But curious; why were these custom MarshalJSON functions needed? Wouldn't our existing code handle the embedded structs just find? Would it still be possible to only produce the bare structs?

corhere replied:
Success! I was able to suppress all the unwanted functions.

thaJeztah commented on api/types/network/network_types.go:
```diff
@@ -0,0 +1,90 @@
+	CheckDuplicate *bool `json:",omitempty"`
+}
+
+// Address represents an IP address
+type Address struct {
+	Addr      string
+	PrefixLen int
+}
```
Got curious why we had this type, but looks like it's only used in fields that are deprecated and those are being removed by @akerouanton in https://github.com/moby/moby/pull/50846

Looks like that PR doesn't remove the `Address` type though; probably should 🤔

thaJeztah commented on api/swagger.yaml:
```diff
@@ -2669,38 +2718,48 @@ definitions:
         description:
           ID of the peer-node in the Swarm cluster.
         type: "string"
+        x-omitempty: false
```
`x-omitempty` is not the default?

thaJeztah replied:
oh! that's surprising; or does it expect all fields that are not "nullable" to be added to `required`?

corhere replied:
Unless the JSON schema says an object property is required, its optional per the JSON Schema Validation spec.

> [An object instance is valid against this keyword if every item in the array is the name of a property in the instance.](https://json-schema.org/draft/2020-12/json-schema-validation#section-6.5.3-2)

Go-swagger is taking that to its logical conclusion: properties that aren't required are only serialized if they are set to a meaningful value. We have the choice to use whichever of `required` or `x-omitempty` is more appropriate for the circumstances.

> The extension does not force a required field to get the “omitempty” modifier.

Given our tendency to deprecate and remove API fields, I think `x-omitempty` is more appropriate than `required` when describing the Engine API [for the same reason required fields are harmful in protobuf land:](https://protobuf.dev/best-practices/dos-donts/#add-required) what's required today may be deprecated tomorrow.

thaJeztah replied:
Yup, that makes sense! And definitely, `x-omitempty` is more suitable to what our intent is. I was wondering if other tooling would treat it different.

thaJeztah reviewed (approved):
LGTM - much better without the weird marshal functions and dependencies

thaJeztah commented:
@akerouanton PTAL - this one will conflict with your other PR; not sure what's the best order to merge; this one first or yours first.

We also need to sync the v1.52 copy of the swagger file in the docs directory; ideally as part of this PR, but otherwise a follow-up is probably fine.
